web3/contract/base_contract.py

```diff
@@ -422,30 +432,40 @@ def __init__(
             self.abi = abi
             self._events = filter_abi_by_type("event", self.abi)
             for event in self._events:
-                setattr(
-                    self,
+                event_factory = contract_event_type.factory(
                     event["name"],
-                    contract_event_type.factory(
-                        event["name"],
-                        w3=w3,
-                        contract_abi=self.abi,
-                        address=address,
-                        event_name=event["name"],
-                    ),
+                    w3=w3,
+                    contract_abi=self.abi,
+                    address=address,
+                    event_name=event["name"],
+                    abi=event,
                 )
+                setattr(self, abi_to_signature(event), event_factory)
 
     def __getattr__(self, event_name: str) -> Type["BaseContractEvent"]:
         if "_events" not in self.__dict__:
             raise NoABIEventsFound(
                 "The abi for this contract contains no event definitions. ",
                 "Are you sure you provided the correct contract abi?",
             )
-        elif event_name not in self.__dict__["_events"]:
+        elif event_name.split("(")[0] not in [
```

Comment:
this `event_name.split...` may deserve to be its own method since you're using it so much in here. Feel free to take or leave :)